### ISSUE TITLE
when increase kube-api-qps param in kube-controller-mananger, etcd2.2.2 out of memory

### ISSUE THREAD

**Issue by magicwang-cn** [kubernetes/kubernetes#18266]
@wojtek-t 
i have created a k8s cluster with 400 nodes, 16000pods[40 pods per node] and use a 5 nodes etcd cluster[with 8G mem, 4 cpus]. and i tested for e2e density, found an etcd problem[OOM], this is the details:

first, definitely, this issue is cased by a param of kube-controller-manager(--kube-api-qps, from 20 to 100, and this can increase the rate of pods add/delete operation)
second, also my doubt, batch creating pods[create a rc with 16000 replicas] will not occur this problem, but batch deleting pods[scale the rc with 0 replicas] does(they both use etcd watch method).
final, it  might be a misuse of etcd?

details:
in the process of deploying 16000 pods will not occur this problem, and the etcd mem use just about 800~900M(total 8G), the etcd snapshort is just about 46M
when i batch deleting pods , the etcd mem use increasing straight up, then OOM.

**Comment by wojtek-t** [comment]
@xiang90 @hongchaodeng @gmarek 

>
first, definitely, this issue is cased by a param of kube-controller-manager(--kube-api-qps, from 20 to 100, and this can increase the rate of pods add/delete operation)

If you are changing this param - then yes it can cause problems. Default values are not set accidentally - those are the values that we think work relatively well.

>
second, also my doubt, batch creating pods[create a rc with 16000 replicas] will not occur this problem, but batch deleting pods[scale the rc with 0 replicas] does(they both use etcd watch method).
final, it might be a misuse of etcd?

Sorry - I'm afraid I didn't understand this comment. Can you please clarify?

**Comment by magicwang-cn** [comment]
@wojtek-t 
thanks for you help.
I mean that the param is effective to both creating and deleting pods, why in the process of deploying 16000 will not occur this problem, but "kubectl scale --replicas=0“ does? What's causing this problem?

**Comment by wojtek-t** [comment]
@magicwang-cn - hmm... I'm not sure. 

>
details:
in the process of deploying 16000 pods will not occur this problem, and the etcd mem use just about 800~900M(total 8G), the etcd snapshort is just about 46M
when i batch deleting pods , the etcd mem use increasing straight up, then OOM.

@xiang90 @hongchaodeng - are folks working also on etcd - any thoughts on it?

**Comment by gmarek** [comment]
During my tests I also saw growth in the etcd resource usage during deletion, though they never were that bad. AFAICT pod addition and deletion are symmetric operation, so there shouldn't be any difference on our part (but I may be missing something @lavalamp @mikedanese). 

To make your setup crystal clear: you're running 5 nodes for etcd + 1 node for master (API server etc.), right? We're generally not running distributed etcd now, so it may be some trait of Raft. Note that I'm shooting in the dark here.

**Comment by magicwang-cn** [comment]
@gmarek @wojtek-t @xiang90
you're running 5 nodes for etcd + 1 node for master (API server etc.), right?--- yes
and this [issue](https://github.com/coreos/etcd/issues/3941) has the etcd log
i do not think it is the trait of Raft, very likely has relation with etcd watch.

**Comment by wojtek-t** [comment]
I don't think it's related to watch, because there is no difference between create & delete "event" from the watch perspective...

**Comment by lavalamp** [comment]
We now do "graceful" deletion by default-- I'm not sure, but that may add some multiplier for etcd reads/writes. @smarterclayton ?

**Comment by smarterclayton** [comment]
Yeah it adds 2-3 writes in the happy path.

Read object (instead of previous blind delete)
Write graceful delete interval, timestamp
Kubelet writes status change (container shutdown)
Kubelet invokes hard delete

Depending on how Kubelet shuts down it can do multiple status writes.

> We now do "graceful" deletion by default-- I'm not sure, but that may add some multiplier for etcd reads/writes.

**Comment by smarterclayton** [comment]
Also, if you create and delete pods rapidly, you may incur a high level of conflict amplification (since initial pod create is followed by several writes from scheduler and Kubelet).  Multiple clients are writing and retrying in that window.

**Comment by lavalamp** [comment]
Thanks, to me that sounds sufficient to explain the reported difference between pod creates and deletes.

**Comment by magicwang-cn** [comment]
@smarterclayton @lavalamp 
so compare to creating pods, deleting pods have more of the following steps with multipier:
[read obj before delete](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/generic/etcd/etcd.go#L382)
[graceful delete](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/generic/etcd/etcd.go#L397-L432)
[status change](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L1922)
am i right?

**Comment by lavalamp** [comment]
Right. So the maximum safe deletion rate may be slower than the maximum
safe creation rate.

**Comment by magicwang-cn** [comment]
@lavalamp @smarterclayton @xiang90 
so is it necessary to seperate creating rate and deleting rate?

**Comment by lavalamp** [comment]
It's a good question. I don't think it's worth the complexity cost at the
moment. Ideally we would want to make an analogous change in all
controllers, not just replication controller, which to me says apiserver
should itself do the rate limiting and push that back onto clients with
429s.

**Comment by Random-Liu** [comment]
@yujuhong Is this related to what we discussed before about the graceful deletion implementation on Kubelet side?

**Comment by yujuhong** [comment]
@Random-Liu, there are suboptimal logic in kubelet's status update that could cause extra GETs after a pod has been deleted, leading to even slower batch deletion. However, I don't think it'd affect the etcd memory usage.

**Comment by magicwang-cn** [comment]
@lavalamp @smarterclayton @hongchaodeng @xiang90 
since i have a look at etcd's code, i found that once etcd receives a request, it will clone the store's data, then write the data into response, @hongchaodeng am i right? [see the code](https://github.com/coreos/etcd/blob/master/etcdserver/etcdhttp/client.go#L669)
so as @smarterclayton said, when deleting,  it adds 2-3 writes in the happy path. the clone data must be large,  this maybe the cause.
![etcd_mem](https://cloud.githubusercontent.com/assets/7587313/12138088/00913a2a-b491-11e5-9d23-215e0eefd704.png)

**Comment by magicwang-cn** [comment]
what's more, when begin to deleting, the prometheus data of etcd is as follows(`with a high growth`):
![etcd_goroutines](https://cloud.githubusercontent.com/assets/7587313/12140095/15b23030-b4a0-11e5-9249-7ac83c07b8b6.png)
![etcd_mem_alloc](https://cloud.githubusercontent.com/assets/7587313/12140099/15e4ff6a-b4a0-11e5-89dc-b028168f9c32.png)
![etcd_mem_heap_inuse](https://cloud.githubusercontent.com/assets/7587313/12140100/15e7152a-b4a0-11e5-924c-828d46708023.png)
![etcd_mem_malloc](https://cloud.githubusercontent.com/assets/7587313/12140098/15e42c8e-b4a0-11e5-91fe-f2c17d851927.png)
![etcd_resident_mem](https://cloud.githubusercontent.com/assets/7587313/12140096/15b5c876-b4a0-11e5-8a4b-a97c0437d8bf.png)
![etcd_whole](https://cloud.githubusercontent.com/assets/7587313/12140097/15b6bbc8-b4a0-11e5-89f5-195c398629a1.png)

**Comment by lavalamp** [comment]
Thanks for the charts. Judging from the stats, it looks like deletes are
not the culprit, since there were only 1000 of them. There's almost 200k
compare&swaps. So I think our "flurry of writes right before delete" theory
is probably correct. That, or deletes are REALLY expensive.

**Comment by xiang90** [comment]
@magicwang-cn Can you also get the store operation (reads, writes, delete, etc..) rate vs time graph? Then we can probably know which operation causes the spike usage.

**Comment by xiang90** [comment]
And it seems the time scale of the graphs are not the same.

**Comment by magicwang-cn** [comment]
OK, i will do this as soon as possible
>@magicwang-cn Can you also get the store operation (reads, writes, delete, etc..) rate vs time graph? Then we can probably know which operation causes the spike usage.

no, they are the data for the same period
>And it seems the time scale of the graphs are not the same.

@xiang90 @lavalamp

**Comment by magicwang-cn** [comment]
@xiang90 @lavalamp @hongchaodeng @smarterclayton @HardySimpson 
this is the store operation graph:
![etcd_reads_total](https://cloud.githubusercontent.com/assets/7587313/12160652/12938758-b529-11e5-9260-ea4a55382ea9.png)
![etcd_watch_request](https://cloud.githubusercontent.com/assets/7587313/12160655/12992cc6-b529-11e5-91f1-e3da64069f91.png)
![etcd_writes_compareandswap](https://cloud.githubusercontent.com/assets/7587313/12160653/1296064a-b529-11e5-97ea-a1c7bc6c996e.png)
![etcd_writes_create](https://cloud.githubusercontent.com/assets/7587313/12160654/12964614-b529-11e5-8e40-d19cb41958fe.png)
![etcd_writes_delete](https://cloud.githubusercontent.com/assets/7587313/12160657/131ff9fe-b529-11e5-883e-561a5788c4fc.png)
![etcd_writes_set](https://cloud.githubusercontent.com/assets/7587313/12160656/129c90dc-b529-11e5-9615-635bc86db200.png)

**Comment by xiang90** [comment]
@magicwang-cn It seems like the 1000 concurrent deletes was killing etcd. Can you find a way to rate limit it and give it another try? Or can you try to give more memory to etcd to see if it survives?

**Comment by magicwang-cn** [comment]
@xiang90 
yeah, apiserver's rate limit may solve this problem, i am planning to have a test(`change the --kube-api-qps param of controller-manager from 20 to 10`).
but every request will do a clone operation may not be a good way? or etcd itself can do the rate limit to keep surviving?

**Comment by magicwang-cn** [comment]
@xiang90 @lavalamp 
first method(`change the --kube-api-qps param of controller-manager from 20 to 10`), the etcd problem still exist.

**Comment by xiang90** [comment]
@magicwang-cn If you look at the graph you generated, the delete rate is apparently much higher than 20 or 10. It is at the order of 100.

**Comment by magicwang-cn** [comment]
no, just about 10/s , you can see as follows:
![etcd_delete_begin](https://cloud.githubusercontent.com/assets/7587313/12164779/41ea9184-b553-11e5-8928-ac17aadb509b.png)
![etcd_delete_end](https://cloud.githubusercontent.com/assets/7587313/12164780/4231dfda-b553-11e5-8462-c2ff2187f89d.png)

>@magicwang-cn If you look at the graph you generated, the delete rate is apparently much higher than 20 or 10. It is at the order of 100.

**Comment by xiang90** [comment]
@magicwang-cn It is not 10 or 20. You showed me the wrong data I think. The slope of pic2 is 10; the slope of pic5 (the delete one I mentioned) is much higher than it.

**Comment by xiang90** [comment]
@magicwang-cn OK. I took a closer look at the pic. The scales of the pics are quite different. But can you show me the rate of etcd_store_writes_total-delete?

**Comment by magicwang-cn** [comment]
@xiang no, pic5 is about 20/s, it's the right data. see as follows:
![etcd_delete_1_end](https://cloud.githubusercontent.com/assets/7587313/12165066/9602ca82-b555-11e5-98ea-b6f76594579f.png)
![etcd_delete_6_begin](https://cloud.githubusercontent.com/assets/7587313/12165065/95d4cf1a-b555-11e5-8182-a55a9eac0eda.png)

**Comment by xiang90** [comment]
@magicwang-cn How did you know the rate from the graph?

**Comment by xiang90** [comment]
@magicwang-cn OK. I see. You calculate the rate by using the data (time, total) showed in the two small dialog.

**Comment by magicwang-cn** [comment]
@xiang90 
the pic1 is: 17:56:19 with 682, and pic2 is: 17:55:57 with 324, so the rate is as follows:
```
(682-324)/22 = 16.3
```

>@xiang no, pic5 is about 20/s, it's the right data. see as follows:
![etcd_delete_1_end](https://cloud.githubusercontent.com/assets/7587313/12165066/9602ca82-b555-11e5-98ea-b6f76594579f.png)
![etcd_delete_6_begin](https://cloud.githubusercontent.com/assets/7587313/12165065/95d4cf1a-b555-11e5-8182-a55a9eac0eda.png)

**Comment by magicwang-cn** [comment]
yeah :)
>@magicwang-cn OK. I see. You calculate the rate by using the data (time, total) showed in the two small dialog.

**Comment by magicwang-cn** [comment]
@xiang90 
[the delete operation](https://github.com/coreos/etcd/blob/master/etcdserver/server.go#L1053) has not spend much memory, as @smarterclayton said, when deleting, there is still some reads and writes operation,  and plus the [clone operation](https://github.com/coreos/etcd/blob/master/etcdserver/etcdhttp/client.go#L669) in etcd, then the problem happen?

>Yeah it adds 2-3 writes in the happy path.

Read object (instead of previous blind delete)
Write graceful delete interval, timestamp
Kubelet writes status change (container shutdown)
Kubelet invokes hard delete

Depending on how Kubelet shuts down it can do multiple status writes.

**Comment by magicwang-cn** [comment]
@davidopp @lavalamp @wojtek-t @kevin-wangzefeng

**Comment by magicwang-cn** [comment]
@xiang90 @lavalamp 
second method(`increate etcd nodes's cpu and memory, with 8 cpu and 16G mem`), the etcd problem still exist.

**Comment by lavalamp** [comment]
@smarterclayton is graceful deletion flag-disableable? I think it'd be a good test to see if the added writes are really causing the problem here or not.

**Comment by magicwang-cn** [comment]
@lavalamp @xiang90 
but sometimes deploying pods will also has this problem, so i think we should find the root cause.

**Comment by magicwang-cn** [comment]
@lavalamp @xiang90 @smarterclayton @hongchaodeng 
i have a test without graceful delete, then the problem is not reproduce.
to sum up:
- in condition of large nodes and pods(`about 500 nodes 15000 pods`), graceful deleting(`additional reads and writes`)  will case etcd OOM problem, and the root case maybe >the delete operation has not spend much memory, as @smarterclayton said, when deleting, there is still some reads and writes operation, and plus the clone operation in etcd, then the problem happen?
- etcd is not allowed with high operating rate (`for example, increate the --kube-api-qps param of controller-manager `)

what is the consider of graceful delete?

**Comment by xiang90** [comment]
@magicwang-cn 

> etcd is not allowed with high operating rate 

Have you tried to reproduce the OOM issue without k8s. Like creating 100K keys and removing them at high rate. Will you also observe the same memory increase?

**Comment by lavalamp** [comment]
@magicwang-cn 

> i have a test without graceful delete, then the problem is not reproduce.

Thank you so much for doing that.

@xiang90 

> Like creating 100K keys and removing them at high rate. Will you also observe the same memory increase?

I predict this would not be the issue-- I think it's the flurry of writes before the ultimate delete. "graceful delete" in our system doesn't do an immediate delete; it puts a pod into a DELETING state and then waits for kubelet to confirm that all of its containers have stopped (all of which are writes) before actually deleting the pod.

I think the only real way to fix this is server-side rate limiting of *something*. Client-side rate limiting in controller manager doesn't work because the massive number of writes is coming from the distributed hoard of kubelets.

At first I was thinking to rate limit deletes, but really it's the fan-out of writes that causes the problem (otherwise, turning off graceful delete would NOT have fixed the problem). So maybe we need to have a max-in-flight limit for writes.

@davidopp @wojtek-t This is a scalability issue.

**Comment by magicwang-cn** [comment]
@lavalamp
yeah,i agree with you.because in the process of deploying pods,sometimes the problem also exist('especial multi rc'). but with the rate limiting of service side,it will decrease k8s's deploying rate,right?

>I think the only real way to fix this is server-side rate limiting of something

**Comment by magicwang-cn** [comment]
@lavalamp @davidopp @xiang90 
any plan?

**Comment by smarterclayton** [comment]
You can test by deleting all your pods with TerminationGracePeriodSeconds =
0, which results in only a single delete

**Comment by magicwang-cn** [comment]
@smarterclayton i have tested without graceful delete.
and what's more, in the process of deploying(`creating, scheduling and running`) pods will alse has this problem(`about 500 nodes, 12000~15000 pods`).

**Comment by magicwang-cn** [comment]
@lavalamp @smarterclayton @xiang90 @HardySimpson
i think i find the root cause:
**its the list-watch mechanism in k8s that cause the etcd problem.**
there has a watch cache in kube-apiserver, default cache is 1000. and k8s use resourceVersion to keep the top 1000 newest version resources(`pod, node and service.. `),  so if the pods is larger(`maybe more than 12000`), the watch cache can not keep a wide range version resources, the watch request from nodes(`about 500`) will be error(`too old resource version`), then it must list the pods again('this will waste much memory in etcd'). above all, with a mount of about 500 nodes's pods list request(`frequently`), then the etcd OOM problem happen. 

so increase the capability of watch cache in kube-apiserver can reduce the rate of pods list  request, thus the problem can be avoided.

so with a good use of etcd, we should use it graceful in clients or keep rate limit in server, any ideas?

**Comment by xiang90** [comment]
@magicwang-cn Why there is no significant increase of read rate showed on the graph you pulled from prometheus?

**Comment by hongchaodeng** [comment]
@magicwang-cn 

> its the list-watch mechanism in k8s that cause the etcd problem.

That sounds interesting. Can you share any evidence you found? Any traces, logs, or profs?

**Comment by gmarek** [comment]
@magicwang-cn - your theory is interesting. cc @wojtek-t to see if he has some thoughts. I'm not convinced it is the case, because we're not running into this problem in our 1000 node clusters. Indeed at some point we saw 'too old resource version' errors but it stopped before we managed to find a root cause. We do expose a lot of prometheus metrics for cache-hits, and various etcd ones, so if there's a huge number of relists in your cluster it should be reflected in them.

**Comment by magicwang-cn** [comment]
@xiang90 
because the increase of read keeps not long duration, the problem happen

>@magicwang-cn Why there is no significant increase of read rate showed on the graph you pulled from prometheus?

@xiang90 @gmarek @davidopp @lavalamp 
**etcd graph:**
![etcd_13_delete](https://cloud.githubusercontent.com/assets/7587313/12288121/536a95e0-ba0c-11e5-92e3-6b05a113c499.png)
![etcd_13_reads](https://cloud.githubusercontent.com/assets/7587313/12288122/536b3e6e-ba0c-11e5-9a80-480ed717f374.png)
![etcd_13_watch](https://cloud.githubusercontent.com/assets/7587313/12288120/536532b2-ba0c-11e5-8269-7e13d4edb7ad.png)

**kube-apiserver list watch graph:**
![apiserver_listwatch_begin](https://cloud.githubusercontent.com/assets/7587313/12288130/690fea9e-ba0c-11e5-8cd2-a4299ccdb2e0.png)
![apiserver_listwatch_end](https://cloud.githubusercontent.com/assets/7587313/12288133/694b3446-ba0c-11e5-90a3-7ccab818f6d4.png)


@gmarek 1000 nodes has no problem? what's the version of k8s and etcd?

**Comment by gmarek** [comment]
@magicwang-cn version from head with etcd v2.2.1. We're running only density tests though, so there's not a lot of deletes in there. Running load.go test on a big cluster is on my todo list for quite some time...

**Comment by magicwang-cn** [comment]
@gmarek 
hmm.. i am also just running density tests(`with etcd v2.2.2`), when the test is over, it will scale rc's Replicas=0, then the deleting happens. what's more, sometimes in the process of deploying pods(`create-> schedule->running`) will also has this problem.

**Comment by gmarek** [comment]
We're not scaling RC's - we just delete namespace, which in turn deletes everything under it. The error we were seeing was during deployment.

I'd be great if you could run experiment with increased cache size, so we would be able to compare results.

**Comment by magicwang-cn** [comment]
OK, i know. i have increase pods's watch size from 1000 to 10000, and the problem is avoided. and the apiserver's max memory use is up to ~28%(`8 cpu and 16G`)

**Comment by gmarek** [comment]
Looks good. If you send a PR I can review it.

**Comment by magicwang-cn** [comment]
is it necessary to make the watch cache size as a param of kube-apiserver? for example(`--watch-cache-sizes=[pods#10000,nodes#1000]`)

**Comment by gmarek** [comment]
It'd be good, as in small clusters with small master machines we probably can't spare memory to keep 10k pods.

**Comment by magicwang-cn** [comment]
well. l will send a PR later.

**Comment by gmarek** [comment]
Thanks!

**Comment by lavalamp** [comment]
Excess lists is possibly a symptom of some component not keeping up with a watch. E.g., at one point there was a queue of watch results in apiserver that would back up. The fact that increasing the watch window size works, implies to me that perhaps kubelet has a similar problem.

**Comment by gmarek** [comment]
cc @kubernetes/sig-scalability

**Comment by timothysc** [comment]
please label area-performance

**Comment by hongchaodeng** [comment]
IMHO increasing cache size is not a scalable option.
I think there is fundamental problem in the current implementation that we should fix.
Here's my analysis:
- 

BTW, client

**Comment by hongchaodeng** [comment]
IMHO increasing cache size is not a scalable option.
I think there is fundamental problem in the current implementation that we should fix.
Here's my analysis:

In [cacher.Watch](https://github.com/kubernetes/kubernetes/blob/60fc2bc09e0fad9a41bd027f9d59abcd10f937d8/pkg/storage/cacher.go#L251-L251), if the version is too old and not in the cache, it just returns error.
```Go
	initEvents, err := c.watchCache.GetAllEventsSinceThreadUnsafe(watchRV)
	if err != nil {
		return nil, err
	}
```
I think the correct behavior should be watch on etcd instead of getting from cache in such case. It would be a network call, but nonetheless since it's old version it should get the result non-blockingly.

/cc @xiang90 @gmarek @timothysc  @wojtek-t @lavalamp

**Comment by magicwang-cn** [comment]
+1

**Comment by gmarek** [comment]
@hongchaodeng - I'm not sure if it's not scalable enough. Memory is cheap, and having, say 5000 Nodes times 100 pods/node gives ~500k entries. Storing everything would probably take small few GB of memory for a huge cluster (assuming reasonable format of storing data, not JSON). Currently we're really low on memory usage on master side, and I'm not sure it's worth optimizing for it. I think currently we're happy to trade memory for CPU in master components - @wojtek-t.

Having said that - I like where @hongchaodeng is going, but I'd like to see some more details.

**Comment by smarterclayton** [comment]
Memory in etcd and master server will continue to be optimized through protobuf over the next few releases, so we'll be reducing a significant amount of I/O and garbage.  Etcd3 should also reduce total memory use.

**Comment by timothysc** [comment]
@hongchaodeng @magicwang-cn you should see the error in the logs, is the log data available?

**Comment by magicwang-cn** [comment]
@smarterclayton already in the plan?

@timothysc the log is (`too old resourceVersion`)

**Comment by smarterclayton** [comment]
It'll be 1.3 at this point, merely experimental until then

**Comment by hongchaodeng** [comment]
@magicwang-cn showed me the trace and it seems like kubelet as client is running Reflector.ListAndWatch() which is causing the flood -- everytime Watch returns an error, it reList.

> if your theory is correct, we should be able to repro with a UT on cache layer.

@timothysc That should be our goal. Now the flow is pretty clear: Reflector -> (Cacher, Reflector, watchCache). If my thought is right, it should not be hard to reproduce.

The hard part is how to measure memory?

**Comment by lavalamp** [comment]
@hongchaodeng 

> Anyone could help explain why the client needs to relist if the watch returns error? It sounds to me that client is behaving too aggressive.

If a watch returns an error, that means changes have been made since the client listed. If the client needs to have a 100% correct state, it must then relist. Otherwise, it may never return to a correct state (imagine one of the missed changes was a deletion).

A flood of lists is quite possibly a cause of the problem, but it's *not the ultimate cause*. The question is **why did the previous watch close**? Whatever causes watches to close prematurely is the thing that needs to be fixed.

**Comment by hongchaodeng** [comment]
@lavalamp 
See [my analysis above](https://github.com/kubernetes/kubernetes/issues/18266#issuecomment-172737668).

There are three identities:
- **client** with a reflector reading states.
- **cacher** with a reflector updating watch cache
- **watch cache** inside cacher

Here's detailed error analysis:
- Watch cache holds fixed size of a queue of data.
- Watch cache is update really fast because there are a lot more state changes than it can hold
- Cacher returns error in Watch() because watch cache didn't contain the old version data.
- Client watch is closed because cacher returns an error (too old resource version).

**Comment by hongchaodeng** [comment]
FYI, I'm trying to reproduce it in UT.

**Comment by lavalamp** [comment]
@hongchaodeng I was responding to your analysis.

There's two cases.

a) a list has just completed and we're attempting to establish a watch.
b) a watch has been initiated and we're trying to keep up.

The situation you're outlining is a potential problem for case a), when we're trying to get a watch started. But that means we've just done a list for some reason-- why did we need to do that list at all? My claim is that something must have gone wrong in steady state, b).

**Comment by hongchaodeng** [comment]
@gmarek @timothysc

I talked with @lavalamp today w.r.t. this issue. Let me give a summary:
- Daniel is worried that the cluster state has been somehow unsteady. That's the point that we should figure out and maintain it steady.
- Daniel pointed out that even though the cache size has a limit, current etcd v2 also has a limit on the history of queue. Fortunately, etcd v3 has a theoretically infinite list (LSM tree) and different semantics. So I would research this topic further and give more analysis.
- Increasing cache size should be a quick duct tape. @magicwang-cn told me he's OK with that.

**Comment by hongchaodeng** [comment]
At the mean time, I could verify:
- the Watch() error of too old resource version could happen 
- List() goes directly to etcd happened [here]( https://github.com/kubernetes/kubernetes/blob/60fc2bc09e0fad9a41bd027f9d59abcd10f937d8/pkg/storage/cacher.go#L284) for a number of resources like "services/specs", "pods", "minions", "nodes".

I still think that the outdated version of watch leads to such problematic unsteady state. Unfortunately I couldn't find where the client logic. I will continue to research on potential clients and have whole-story analysis of client logic as well.

**Comment by lavalamp** [comment]
List of pods should be served out of the cache. The line you linked to is
for clients that don't set the resource version; our own components ought
to be setting the resource version.

**Comment by magicwang-cn** [comment]
@lavalamp yes, i see that even though the Watch() failed with (`too old resourceVersion`), if the resourceVersion is setted, the List() also reads data from cache. but from the graph i have intercepted before, when deleting pods, the list action has an obviously increasing.
case by watch list [itself](https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/cacher.go#L184)?

**Comment by lavalamp** [comment]
@magicwang-cn I'm sorry, can you rephrase? I don't understand

**Comment by magicwang-cn** [comment]
@lavalamp i mean if the resource version is [setted](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/cache/reflector.go#L236), even though the watch() [failed with too old resource version](https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/cacher.go#L251-L254), the list() still [reads data from cache](https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/cacher.go#L315)
but from the graph i have got before, when deleting pods, the list() action has an obviously increasing(`see as follows`):

>**kube-apiserver list watch graph:**
![apiserver_listwatch_begin](https://cloud.githubusercontent.com/assets/7587313/12288130/690fea9e-ba0c-11e5-8cd2-a4299ccdb2e0.png)
![apiserver_listwatch_end](https://cloud.githubusercontent.com/assets/7587313/12288133/694b3446-ba0c-11e5-90a3-7ccab818f6d4.png)

**Comment by lavalamp** [comment]
@magicwang-cn I still don't think I understand what you're saying. Note that the list handler returns a different resource version than it gets-- if you send "0", it sends back `readResourceVersion`, which should be the resource version the cache is actually at. If you then do a watch from `readResourceVersion`, you should be in the window.

The charts with a huge spike in lists tells me that there was a correlated watch failure across multiple components. The lists themselves are a symptom of the underlying cause.

**Comment by timothysc** [comment]
@magicwang-cn what objects are you failing on in the cache?  Is it consistent? 

xref: https://github.com/kubernetes/kubernetes/pull/19970

**Comment by magicwang-cn** [comment]
@lavalamp @hongchaodeng @timothysc 
today i have found the potential clients that send the List() request without set (`resourceVersion=0`), mainly in density test itself(`especial DeleteRC, this will list pods from etcd with a long time `), for example:
[DeleteRC](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/util.go#L1921)
[start](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/util.go#L1694)

and i have  printed the log that behind [this line](https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/cacher.go#L283), find there are still many clients use List() request without set resourceVersion.

so what's the 
consider of [this judgment](https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/cacher.go#L281-#L285)

**Comment by lavalamp** [comment]
It would be useful to log the client there so we can get a list of which
components need to be updated.

I think only clients on every node really matter-- I don't think the e2e
tests themselves will make enough traffic to harm the server, but I still
think it'd be good to fix them.

**Comment by timothysc** [comment]
> It would be useful to log the client there so we can get a list of which components need to be updated.

+1 , also I wouldn't mind a V(4) log on above to diagnose.

**Comment by wojtek-t** [comment]
Sorry for jumping late into discussion - I was on paternity leave for a month.

Generally, I kind of agree with Daniel, that if increasing the size of cache helps, we have much more serious problem and we shouldn't fix it by changing cache sizes.

>
- Watch cache holds fixed size of a queue of data.
- Watch cache is update really fast because there are a lot more state changes than it can hold
- Cacher returns error in Watch() because watch cache didn't contain the old version data.
- Client watch is closed because cacher returns an error (too old resource version).
- Client relists after watch is closed and it causes flood to etcd

The question is exactly why "Cacher" returns error in Watch - it definitely shouldn't.
Basically, watch event is emitted to watchers synchronously with reading it from etcd. So this should never happen.
I'm happy to help with debugging this, but I need a better way for reproducing it.

**Comment by spzala** [comment]
@magicwang-cn hello, since this issue is over a year old and also etcd 3.2 is out - can we please verify if this issue is still valid? If not, can we please close it? Related issue #20540 has a comment that etcd3 should solve problems related to large cluster. Thanks!
